src/EditorFeatures/CSharpTest/CodeActions/ConvertLinq/ConvertLinqQueryToLinqMethodTests.cs

```diff
@@ -0,0 +1,151 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.CodeRefactorings;
+using Microsoft.CodeAnalysis.ConvertLinq;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.CodeActions.ConvertLinq
+{
+    public class ConvertLinqQueryToLinqMethodTests : AbstractCSharpCodeActionTest
+    {
+        protected override CodeRefactoringProvider CreateCodeRefactoringProvider(Workspace workspace, TestParameters parameters)
+            => new CodeAnalysis.CSharp.ConvertLinq.CSharpConvertLinqQueryToLinqMethodProvider();
+
+        #region Diagnostics
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertLinq)]
+        public async Task Conversion_WhereOrderByTrivialSelect()
+        {
+            await Test(
+@"from num in new int[] { 0, 1, 2 }
+where num % 2 == 0
+orderby num
+select num",
+@"new int[] { 0, 1, 2 }.Where(num => num % 2 == 0).OrderBy(num => num)");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertLinq)]
+        public async Task Conversion_WhereSelect()
+        {
+            await Test(
+@"from x in new int[] { 0, 1, 2 } where x< 5 select x* x",
+@"new int[] { 0, 1, 2 }.Where(x => x < 5).Select(x => x * x)");
```

Comment:
it looks like the rewrite is reformatting the expressions.  Is that what we want?

Reply:
Now we get the lambda syntax from IOperation nodes. It seems that it is already formatted there. @mavasani, @alekseyts to confirm if we can get unmodified syntax.

---
In reply to: [167387460](https://github.com/dotnet/roslyn/pull/24759#discussion_r167387460) [](ancestors = 167387460)

Reply:
> Now we get the lambda syntax from IOperation nodes. It seems that it is already formatted there.

I am not sure what does "formatted" means in this context. IOperation doesn't format anything, it references some original syntax nodes. Also, I would not try to rely on syntax associated with IOperation nodes that represent compiler synthesized things (like synthesized lambdas). If the node is not synthesized, its syntax is exactly the one that was used to find it. So, I would recommend to always work of off the original syntax tree, rather than of off IOperation.Syntax.

Reply:
Just found that we add/remove whitespaces as a postprocess here:
[http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/CodeActions/CodeAction.cs,275](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/CodeActions/CodeAction.cs,275)

I'm not sure we should avoid this. Thoughts?

---
In reply to: [171135570](https://github.com/dotnet/roslyn/pull/24759#discussion_r171135570) [](ancestors = 171135570,167387460)

Reply:
We should only be formatting code that has elastic annotations on it.  Normal user code should not have those.  In this case, your lambda probably has these annotations on it, so the formatter s going and fixing htings up.

I'm not sure the best way to deal with this.  @heejaechang might know.  If there is a DoNotFormat annotation, we might want to add that to the parts of the code that came from the original query.

Reply:
if you mix existing code with new node, new tree will contains a marker that tells formatter to format. here, formatter might or might not do the formatting as you wish since usually those are different case by case.

if you don't want formatter to kick in, make sure you give trivia explicitly when you create new node.

Reply:
Tried to keep trivias. did not get them in correct locations but preserved in locations we should not keep (end lines).  Added this feature to [https://github.com/dotnet/roslyn/issues/25112](https://github.com/dotnet/roslyn/issues/25112)

---
In reply to: [171443318](https://github.com/dotnet/roslyn/pull/24759#discussion_r171443318) [](ancestors = 171443318)